### PR TITLE
Asset pipeline refactor

### PR DESCRIPTION
**Pull request by YahimaB**

### CODE REVIEW

File: Engine/BaseAssets/Components/Transform.cs [hunk 1]
```diff
@@ -1,32 +1,170 @@
...
+using System.Collections.Immutable;
+using System.Collections.ObjectModel;
 using System.Reflection;
-
 using LinearAlgebra;
 
 namespace Engine.BaseAssets.Components
 {
-    public class Transform : Component, INotifyFieldChanged
+    public partial class Transform : Component, INotifyFieldChanged
```
Comment (Notrum666) [inline Engine/BaseAssets/Components/Transform.cs:10]: remove partial declarations, use regions if needed

File: Engine/GameObject.cs [hunk 2]
```diff
@@ -1,126 +1,158 @@
...
+        private readonly List<Component> components = new List<Component>();
+
+        public string Name { get => name; set => name = value; }
+        public bool LocalEnabled { get => localEnable; set => localEnable = value; }
+        public Transform Transform { get => transform; private init => transform = value; }
         public ReadOnlyCollection<Component> Components => components.AsReadOnly();
-        internal bool PendingDestroy { get; private set; }
-        public bool Initialized { get; internal set; }
+
+        public bool Enabled => localEnable && Transform != null && (Transform.Parent?.GameObject?.Enabled ?? true); //TODO: maybe better cache value
```
Comment (Notrum666) [inline Engine/GameObject.cs:25]: use "is not null" instead of !=

File: Engine/GameObject.cs [hunk 3]
```diff
@@ -1,126 +1,158 @@
...
+    public sealed class GameObject : SerializableObject
     {
-        private bool enabled = true;
-        public bool Enabled { get => Transform.Parent == null ? enabled : enabled && Transform.Parent.GameObject.enabled; set => enabled = value; }
-        public Transform Transform { get; private set; }
-        private List<Component> components = new List<Component>();
+        [SerializedField]
+        private string name;
+        [SerializedField]
+        private bool localEnable = true;
```
Comment (Notrum666) [inline Engine/GameObject.cs:14]: typo - localEnabled

File: Engine/GameObject.cs [hunk 4]
```diff
@@ -1,126 +1,158 @@
...
+        [SerializedField]
+        private Transform transform = null;
+        [SerializedField]
+        private readonly List<Component> components = new List<Component>();
+
+        public string Name { get => name; set => name = value; }
```
Comment (Notrum666) [inline Engine/GameObject.cs:20]: make sure name is not null or empty/whitespace

File: Engine/GameObject.cs [hunk 5]
```diff
@@ -1,126 +1,158 @@
...
+        public bool LocalEnabled { get => localEnable; set => localEnable = value; }
+        public Transform Transform { get => transform; private init => transform = value; }
         public ReadOnlyCollection<Component> Components => components.AsReadOnly();
-        internal bool PendingDestroy { get; private set; }
-        public bool Initialized { get; internal set; }
+
+        public bool Enabled => localEnable && Transform != null && (Transform.Parent?.GameObject?.Enabled ?? true); //TODO: maybe better cache value
 
         public GameObject()
         {
```
Comment (Notrum666) [inline Engine/GameObject.cs:28]: generate some name

File: Engine/GraphicsCore.cs [hunk 6]
```diff
@@ -240,90 +204,80 @@ public static void Update()
...
 
-            List<GameObject> objects = EngineCore.CurrentScene.objects;
+            // List<GameObject> objects = EngineCore.CurrentScene.Objects;
 
             ShaderPipeline pipeline = null;
 
             void renderObjects()
             {
-                foreach (GameObject obj in objects)
+                foreach (MeshComponent meshComponent in Scene.FindComponentsOfType<MeshComponent>())
```
Comment (Notrum666) [inline Engine/GraphicsCore.cs:219]: add todo to test performance of manual objects iteration (allowing skipping of non-localEnabled objects) vs this implementation (that goes up the gameobject tree for every component)

File: SharpDXtest/EditorWindow.xaml.cs [hunk 7]
```diff
@@ -36,9 +37,14 @@ public EditorWindow()
...
         {
+            var mainPath = Directory.GetCurrentDirectory();
+            var solutionPath = Directory.GetParent(mainPath)?.Parent?.Parent?.Parent?.Parent?.FullName;
+            AssetsManager.InitializeInFolder(solutionPath);
+            AssetsRegistry.InitializeInFolder(solutionPath);
+            
             EngineCore.Init(new WindowInteropHelper(this).Handle, (int)ActualWidth, (int)ActualHeight);
 
-            EngineCore.CurrentScene = AssetsManager_Old.LoadScene("Assets\\Scenes\\Scene5.xml");
+            CreateBaseScene();
```
Comment (Notrum666) [inline SharpDXtest/EditorWindow.xaml.cs:47]: please recreate scene5 in a form of base asset and load it instead

**Review by Notrum666 — CHANGES_REQUESTED** [review]

File: Engine/Assets/Scene.cs [hunk 8]
```diff
@@ -19,5 +28,72 @@ protected override void Dispose(bool disposing)
...
+            newObjects.Add(newObject);
+        }
+
+        internal void ProcessNewObjects()
+        {
+            foreach (GameObject newObject in newObjects.ToImmutableArray())
+            {
+                newObject.OnDestroyed += OnObjectDestroyed;
+                gameObjects.Add(newObject);
+                newObjects.Remove(newObject);
```
[the remark this answers is not in the document]
Comment (YahimaB) [reply Engine/Assets/Scene.cs:89]: Leave as protection

File: Engine/BaseAssets/Components/Transform.cs (hunk 1 above)
Comment (YahimaB) [reply Engine/BaseAssets/Components/Transform.cs:10]: booooo

File: Engine/GameObject.cs (hunk 3 above)
Comment (YahimaB) [reply Engine/GameObject.cs:14]: done

File: Engine/GraphicsCore.cs (hunk 6 above)
Comment (YahimaB) [reply Engine/GraphicsCore.cs:219]: Use components LocalEnabled instead of Enabled

File: SharpDXtest/EditorWindow.xaml.cs (hunk 7 above)
Comment (YahimaB) [reply SharpDXtest/EditorWindow.xaml.cs:47]: Another branch - tag components with SerializedField

File: Engine/GameObject.cs (hunk 5 above)
Comment (YahimaB) [reply Engine/GameObject.cs:28]: Done

File: Engine/GameObject.cs (hunk 4 above)
Comment (YahimaB) [reply Engine/GameObject.cs:20]: Done

File: Engine/GameObject.cs (hunk 2 above)
Comment (YahimaB) [reply Engine/GameObject.cs:25]: [https://www.youtube.com/watch?v=o-YBDTqX_ZU&ab_channel=MusRest](Do not check)

**Review by Notrum666 — APPROVED** [review]

**Pull request merged**